### DISCUSSION TITLE
Multiple instances on external site

### DISCUSSION THREAD

**Discussion by cristianvogel** [cables-gl/cables_docs#666] [Answers & Questions] [answered]
Hi!

I am embedding dials designed in Cables for a native WebView based plugin.  Still trying to decide if that is a good idea or not. I like the design flow in Cables, the export to GitHub ( and easy sync via git submodule ) - but wondering if it is practical to have as many instnces of CABLES running, as there are GUI widgets. 

One advantage is that I can use my own layout and other stuff handling and scaling multiple canvas elements running Cables sketches. I have done that and stored each Cables patch instance in a Svelte store. Proof of Concept works. 

Wondering if that is viable in terms of scaling though. 

Should I try to build the whole GUI, including Widget instances and layout in one Cable sketch and export that? Or can I instantiate one CABLES global object target multiple Canvasses running the same Patch code? 

I am guessing this is something Cables dev crew have thought about, considering the move to native dev...

**Comment by steam0r** [comment]
hey,

not really sure how this relates to "native dev" and what you mean by that :) 

having one patch for each element seems like a very weird idea, if you go that way you can use cables-cli (https://github.com/cables-gl/cables-cli) to at least export and use the cables core code only once (via the `-C` switch and some trickery). but even that would mean you have multiple canvases instanciated, right?

wouldnt it be somehow better to have the whole experience in one "widget" and then handle state in cables. not really sure about "svelte" and what exactly "gui widgets" are though.

**Reply by steam0r** [reply]
"trickery": https://github.com/cables-gl/cables_docs/discussions/595

**Reply by cristianvogel** [reply]
Sorry, I will try to clarify. 
A gui widget is a Fader or a Dial. 
Svelte is a front end framework. 
Native Dev , I referred to [recent Cables news](https://www.youtube.com/watch?v=G6QFFDCYmn4 ) about 'offline'
 
 I want to manage state in my app, not in cables. I want to draw things and design interaction in Cables.

**Reply by cristianvogel** [reply]
Also , ```having one patch for each element seems like a very weird idea``` 
Maybe, it is. But it is where I am. Because, I need to instantiate faders and dials dynamically at runtime, so I need to be able to not know how many I am going to need at design stage, hence not being able to design it all in one patch. Is that really so weird?

**Reply by steam0r** [reply]
"I want to manage state in my app, not in cables. I want to draw things and design interaction in Cables."

maybe you could set some variables form the outside, so svelte tells cables when and where to show widgets. or trigger from the outside. then have only one canvas that resides in the background and is "black" most of the time...?

https://dev.cables.gl/docs/4_export_embed/dev_embed_functions/dev_embed_functions

**Reply by cristianvogel** [reply]
Like, with GridTransform.op you mean ? I tried that just now, looks nice, I get loads of 'widgets' neatly arranged. But the mouse interaction then is tricky. How do I know which one is which, and that they send seperate data values to the program? Is there an example?

**Reply by cristianvogel** [reply]
And also, just to know in general and for others to know best practice. Is it not a good pattern to create and manage lots of small canvas elements in front end dev? In Cables editor, there's all those little editor Viz graphs and stuff. Aren't they small canvas elements, independent of each other?

**Reply by steam0r** [reply]
i think firefox will limit you to something like 10-ish webgl canvases, for starters. we hide vizops rendering when they are not visible, even with that sometimes firefox just shows black boxes, if you have a lot of them.

about "gridtransform" that really depends on what kind of things your elements are made of...maybe take a look at intersect ops if you are using 3d https://dev.cables.gl/ops/Ops.Graphics.Intersection

**Reply by cristianvogel** [reply]
This is a cool solution that works, because my 'widgets' are indeed 3D meshes. 
Only thing is that `GridTransform.op` doesn't seem to replicate the `IntersectBody.op` . 
<img width="382" alt="Screenshot 2024-01-31 at 16 00 41" src="https://github.com/cables-gl/cables_docs/assets/12829050/1690ac5f-ffa2-4680-8f8d-10bbc0d30f3e">

**Reply by cristianvogel** [reply]
<img width="385" alt="Screenshot 2024-01-31 at 16 05 07" src="https://github.com/cables-gl/cables_docs/assets/12829050/d1ef63af-0ea4-48a9-8811-57952641ebb9">

If I do a whole seperate branch manually, this works. I still can't trigger the instances from outside though, only because the `IntersectBody.op` doesn't seem to replicate off a `Repeat2D` or `GridTransform` - would be neat if it did.

**Comment by cristianvogel** [comment]
I'm still finding it testing to render an arbitrary number of interactive widgets, like Faders or Dials, and get values from them each independently, using the Grid or Repeat2D . Not sure if I am getting confused about the simple way to do this.

**Comment by cristianvogel** [comment]
@steam0r I invited you and @pandrr to the patch, because there's weirdness that I can't fathom, over in the Data section. Why doesn't the array store the values from the hits? It seems to be creating a new array everytime, so I can't do much in terms of state on each interactive widget.

**Reply by steam0r** [reply]
maybe you find a way to set the positions of the elements xyz and then use meshinstance or something similar, see this part from the "intersections example", intersect body takes positions:

![Bildschirmfoto 2024-01-31 um 17 34 49](https://github.com/cables-gl/cables_docs/assets/151271/17eaac25-6b55-4099-a7fa-58be4ca4ab68)

**Reply by cristianvogel** [reply, marked as answer]
I figured something out. 

Thanks for you help!
